### PR TITLE
Adding displacement BDF2

### PR DESCRIPTION
**Pull request by loumalouomega**
Sorry for the delay, the problem wasn't in the formulation, I am so ret**** that I forgot to set to 3 the buffer (and the check was in 2 because copy and paste from Bossak, sorry again)

I will buy one of this to never forget my stupidity

imagen

### CODE REVIEW

**Comment by rubenzorrilla** [comment]
🎖

File: kratos/solving_strategies/schemes/residual_based_bdf2_displacement_scheme.h [hunk 1]
```diff
@@ -0,0 +1,600 @@
...
+ * - Displacements:
+ *     -# \f$ u_{n0} \f$ the displacement at the current step
+ *     -# \f$ u_{n1} \f$ the displacement one step in the past
+ *     -# \f$ u_{n2} \f$ the displacement two steps in the past
+ * 
+ * Then we assume:
+ *  \f[ \frac{d(vn0)}{dt} \|t_{n0} = c_0 v_{n0} + c_1 v_{n1} + c_2 v_{n2} \f]
+ *  \f[ \frac{d(un0)}{dt} \|t_{n0} = c_0 u_{n0} + c_1 u_{n1} + c_2 u_{n2} \f]
+ * with:
+ *  -# \f$ c_0 = \frac{1.5}{dt} \f$
```
Comment (RiccardoRossi) [inline kratos/solving_strategies/schemes/residual_based_bdf2_displacement_scheme.h:69]: please consider the case of variable time steps. take a look e.g. at the processes/compute_bdfcoefficients_process.h
 which has the relations implemented.

File: kratos/solving_strategies/schemes/residual_based_bdf2_displacement_scheme.h [hunk 2]
```diff
@@ -0,0 +1,600 @@
...
+            if (it_dof->IsFree())
+                it_dof->GetSolutionStepValue() += TSparseSpace::GetValue(Dx,it_dof->EquationId());
+        }
+
+        // Updating time derivatives (nodally for efficiency)
+        const int num_nodes = static_cast<int>(rModelPart.Nodes().size());
+
+        #pragma omp parallel for
+        for(int i = 0;  i < num_nodes; ++i) {
+            auto it_node = rModelPart.Nodes().begin() + i;
```
Comment (RiccardoRossi) [inline kratos/solving_strategies/schemes/residual_based_bdf2_displacement_scheme.h:202]: this is ok, you don't need to change this. 

However just note that you also have the function NodesBegin()

File: kratos/solving_strategies/schemes/residual_based_bdf2_displacement_scheme.h [hunk 3]
```diff
@@ -0,0 +1,600 @@
+//    |  /           |
```
Comment (RiccardoRossi) [inline kratos/solving_strategies/schemes/residual_based_bdf2_displacement_scheme.h:1]: suggestion: why don't you make it to also consider the BDF1 scheme? it is first order accurate but useful if you do want a lot of dissipation, for example in pseudo-dynamic problems

File: kratos/solving_strategies/schemes/residual_based_bdf2_displacement_scheme.h (hunk 1 above)
Comment (RiccardoRossi) [reply kratos/solving_strategies/schemes/residual_based_bdf2_displacement_scheme.h:69]: mmm i think you did implement the variable time step case, however the comment is outdated

File: kratos/solving_strategies/schemes/residual_based_bdf2_displacement_scheme.h [hunk 4]
```diff
@@ -0,0 +1,600 @@
...
+    
+    void InitializeSolutionStep(
+        ModelPart& rModelPart,
+        TSystemMatrixType& A,
+        TSystemVectorType& Dx,
+        TSystemVectorType& b
+        ) override
+    {
+        KRATOS_TRY;
+
```
Comment (RiccardoRossi) [inline kratos/solving_strategies/schemes/residual_based_bdf2_displacement_scheme.h:324]: check the modelpart buffer size here

File: kratos/solving_strategies/schemes/residual_based_bossak_displacement_scheme.hpp [hunk 5]
```diff
@@ -160,84 +145,68 @@ class ResidualBasedBossakDisplacementScheme: public Scheme<TSparseSpace,TDenseSp
...
-            array_1d<double, 3 > DeltaDisplacement;
-
-            NodesArrayType::iterator itNode = NodeBegin + i;
+        const int num_nodes = static_cast<int>(rModelPart.Nodes().size());
 
-            noalias(DeltaDisplacement) = (itNode)->FastGetSolutionStepValue(DISPLACEMENT) - (itNode)->FastGetSolutionStepValue(DISPLACEMENT, 1);
-
-            array_1d<double, 3 > & CurrentVelocity            = (itNode)->FastGetSolutionStepValue(VELOCITY, 0);
-            const array_1d<double, 3 > & PreviousVelocity     = (itNode)->FastGetSolutionStepValue(VELOCITY, 1);
+        #pragma omp parallel for
```
Comment (RiccardoRossi) [inline kratos/solving_strategies/schemes/residual_based_bossak_displacement_scheme.hpp:190]: this is a good change. however WHY IS THIS PART OF THIS PR?

it is a file used by a lot of people so please move this to its own PR

File: kratos/solving_strategies/schemes/residual_based_bossak_displacement_scheme.hpp [hunk 6]
```diff
@@ -899,199 +401,161 @@ class ResidualBasedBossakDisplacementScheme: public Scheme<TSparseSpace,TDenseSp
...
-
-            noalias(RHS_Contribution)  -= prod(M, mVector.a[thread]);
+            noalias(RHS_Contribution) -= prod(M, a);
         }
 
         // Adding damping contribution
         // Damping contribution
-        if (D.size1() != 0)
-        {
-            rCurrentCondition->GetFirstDerivativesVector(mVector.v[thread], 0);
```
Comment (RiccardoRossi) [inline kratos/solving_strategies/schemes/residual_based_bossak_displacement_scheme.hpp:1090]: this was done this way to avoid reallocating. Why did u decide to change this???

**Review by RiccardoRossi — CHANGES_REQUESTED** [review]

File: kratos/solving_strategies/schemes/residual_implicit_time_scheme.h [hunk 7]
```diff
@@ -0,0 +1,530 @@
+//    |  /           |
```
Comment (RiccardoRossi) [inline kratos/solving_strategies/schemes/residual_implicit_time_scheme.h:1]: what is the purpose of this file?

File: kratos/solving_strategies/schemes/residual_based_bdf2_displacement_scheme.h [hunk 8]
```diff
@@ -0,0 +1,600 @@
...
+            const array_1d<double, 3>& vn2 = it_node->FastGetSolutionStepValue(VELOCITY,     2);
+            const array_1d<double, 3>& un2 = it_node->FastGetSolutionStepValue(DISPLACEMENT, 2);
+            const array_1d<double, 3>& an1 = it_node->FastGetSolutionStepValue(ACCELERATION, 1);
+            const array_1d<double, 3>& vn1 = it_node->FastGetSolutionStepValue(VELOCITY,     1);
+            const array_1d<double, 3>& un1 = it_node->FastGetSolutionStepValue(DISPLACEMENT, 1);
+            array_1d<double, 3>& an0 = it_node->FastGetSolutionStepValue(ACCELERATION);
+            array_1d<double, 3>& vn0 = it_node->FastGetSolutionStepValue(VELOCITY);
+            array_1d<double, 3>& un0 = it_node->FastGetSolutionStepValue(DISPLACEMENT);
+
+            if (it_node -> IsFixed(ACCELERATION_X)) {
```
Comment (msandre) [inline kratos/solving_strategies/schemes/residual_based_bdf2_displacement_scheme.h]: What sort of problem would fix acceleration/vel? Should Check() also consider if velocity and acceleration are dofs?

**Review by msandre — DISMISSED** [review]
I don't have any additional comments from my side.

File: kratos/solving_strategies/schemes/residual_based_bdf2_displacement_scheme.h (hunk 3 above)
Comment (loumalouomega) [reply kratos/solving_strategies/schemes/residual_based_bdf2_displacement_scheme.h:1]: I can do it

File: kratos/solving_strategies/schemes/residual_based_bdf2_displacement_scheme.h (hunk 1 above)
Comment (loumalouomega) [reply kratos/solving_strategies/schemes/residual_based_bdf2_displacement_scheme.h:69]: I consider it later

File: kratos/solving_strategies/schemes/residual_based_bdf2_displacement_scheme.h (hunk 4 above)
Comment (loumalouomega) [reply kratos/solving_strategies/schemes/residual_based_bdf2_displacement_scheme.h:324]: I do

File: kratos/solving_strategies/schemes/residual_based_bossak_displacement_scheme.hpp (hunk 5 above)
Comment (loumalouomega) [reply kratos/solving_strategies/schemes/residual_based_bossak_displacement_scheme.hpp:190]: The problem is that I create the implicit_scheme in order to reduce code duplication. I added two tests, one in python and other in cpp, everything is right

File: kratos/solving_strategies/schemes/residual_based_bossak_displacement_scheme.hpp (hunk 6 above)
Comment (loumalouomega) [reply kratos/solving_strategies/schemes/residual_based_bossak_displacement_scheme.hpp:1090]: I will restore it, I didn't know it was because of that

File: kratos/solving_strategies/schemes/residual_based_bdf2_displacement_scheme.h (hunk 8 above)
Comment (loumalouomega) [reply kratos/solving_strategies/schemes/residual_based_bdf2_displacement_scheme.h]: Ok

File: kratos/solving_strategies/schemes/residual_implicit_time_scheme.h (hunk 7 above)
Comment (loumalouomega) [reply kratos/solving_strategies/schemes/residual_implicit_time_scheme.h:1]: The purpose is to reduce code duplication among implicit schemes with this

**Comment by loumalouomega** [comment]
Done, preparing BDF1 now

**Comment by msandre** [comment]
Sorry if I missed it, but to me it is still not clear why we check if vel/accel is fixed. This unusual from my point of view. Could you please include a comment right where you do this to explain the intent/purpose?

**Comment by loumalouomega** [comment]
I close this PR and I create two separates PR

**Comment by RiccardoRossi** [comment]
@msandre the problem is how to allow the user to prescribe the velocity or acceleration. The scheme always solves for DISPLACEMENT, so we must have an indirect mechanism to transform the VELOCITY (or ACCELERATION) input to DISPLACEMENT
this is done by the mechanism you see.

an alternative would be to do it in the preprocessing phase, however the problem with such an approach is that the preprocessor would have to be aware of exactly what scheme is being used in order to make an exact transformation between prescribed VELOCITY and DISPLACEMENT

**Comment by loumalouomega** [comment]
Thanks @RiccardoRossi
